contracts/SafeProtocolMediator.sol

```diff
@@ -0,0 +1,160 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity ^0.8.18;
+import {ISafeProtocolMediator} from "./interfaces/Mediator.sol";
+import {ISafeProtocolModule} from "./interfaces/Components.sol";
+
+import {ISafe} from "./interfaces/Accounts.sol";
+import {SafeProtocolAction, SafeTransaction, SafeRootAccess} from "./DataTypes.sol";
+import {Ownable2Step} from "@openzeppelin/contracts/access/Ownable2Step.sol";
+
+/**
+ * @title TODO
+ * @notice TODO
+ */
+contract SafeProtocolMediator is ISafeProtocolMediator, Ownable2Step {
+    mapping(address => uint) public nonces;
+
+    struct MoudleAccessInfo {
+        bool enabled;
+        bool rootAddressGranted;
+        // TODO: Add deadline for validity
+    }
+
+    /**
+     * @notice Mapping of a mapping what stores information about modules that are enabled per Safe.
+     *         address (Safe address) => address (component address) => EnabledMoudleInfo
+     */
+    mapping(address => mapping(address => MoudleAccessInfo)) public enabledComponents;
+
+    event ActionsExecuted(address safe, bytes32 metaHash);
+    event RootAccessActionsExecuted(address safe, bytes32 metaHash);
+    event ModuleEnabled(address safe, address module, bool allowRootAccess);
+    event ModuleDisabled(address safe, address module);
+
+    error InvalidNonce(address sender, uint256 nonce);
+    error ModuleRequiresRootAccess(address sender);
+    error MoudleNotEnabled(address module);
+    error ModuleEnabledOnlyForRootAccess(address module);
+    error ModuleAccessMismatch(address module, bool requiresRootAccess, bool providedValue);
+
+    constructor(address initalOwner) {
+        _transferOwnership(initalOwner);
+    }
+
+    modifier onlyEnabledModule(ISafe safe) {
+        if (!enabledComponents[address(safe)][msg.sender].enabled) {
+            revert MoudleNotEnabled(msg.sender);
+        }
+        _;
+    }
+
+    /**
+     * @notice TODO
+     * @param safe TODO
+     * @param transaction TODO
+     * @return success TODO
+     * @return data TODO
+     */
+    function executeTransaction(
+        ISafe safe,
+        SafeTransaction calldata transaction
+    ) external override onlyEnabledModule(safe) returns (bool success, bytes[] memory data) {
```

Comment:
Could you outline why you decided to return an array of data over any of the other solutions proposed in the [specs](https://github.com/5afe/safe-protocol-specs/tree/main/mediator)?

Please open a PR to the specs with the proposed interface change, so that we can document why we decided on a specific solution and the discussion around it.

Reply:
My intention here is to return individual result of each of the actions instead of result of last call.
I would comment on this more once this is tested. If all calls success, result of each action will be returned individually back to the caller (i.e component). Downside: additional costs due to gas usage.
Will open PR in specs if it doesn't follow protocol spec in final version.

Reply:
Updated spec: https://github.com/5afe/safe-protocol-specs/pull/4